xmtp_mls/src/storage/sql_key_store.rs

```diff
@@ -22,60 +32,923 @@ impl<'a> SqlKeyStore<'a> {
     pub fn conn(&self) -> &DbConnection<'a> {
         self.conn
     }
+
+    pub fn write<const VERSION: u16>(
+        &mut self,
+        label: &[u8],
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), <Self as StorageProvider<CURRENT_VERSION>>::Error> {
+        let storage_key = build_key_from_vec::<VERSION>(label, key.to_vec());
+        let query = "REPLACE INTO storage (storage_key, version, data) VALUES (?, ?, ?)";
+
+        self.conn.raw_query(|conn| {
+            sql_query(query)
+                .bind::<diesel::sql_types::Binary, _>(&storage_key)
+                .bind::<diesel::sql_types::Text, _>(&VERSION.to_string())
+                .bind::<diesel::sql_types::Binary, _>(&value)
+                .execute(conn)
+        });
+
+        Ok(())
+    }
+
+    pub fn append<const VERSION: u16>(
+        &mut self,
+        label: &[u8],
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), <Self as StorageProvider<CURRENT_VERSION>>::Error> {
+        let storage_key = build_key_from_vec::<VERSION>(label, key.to_vec());
+        let query = "INSERT INTO storage (storage_key, version, data) VALUES (?, ?, JSON_ARRAY_APPEND(COALESCE((SELECT data FROM storage WHERE storage_key = ? AND version = ?), '[]'), '$', ?)) ON DUPLICATE KEY UPDATE data = VALUES(data)";
+        self.conn.raw_query(|conn| {
+            sql_query(query)
+                .bind::<diesel::sql_types::Binary, _>(&storage_key)
+                .bind::<diesel::sql_types::Text, _>(&VERSION.to_string())
+                .bind::<diesel::sql_types::Binary, _>(&storage_key)
+                .bind::<diesel::sql_types::Text, _>(&VERSION.to_string())
+                .bind::<diesel::sql_types::Binary, _>(&value)
+                .execute(conn)
+        });
+        Ok(())
+    }
+
+    pub fn remove_item<const VERSION: u16>(
+        &mut self,
+        label: &[u8],
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), <Self as StorageProvider<CURRENT_VERSION>>::Error> {
+        let storage_key = build_key_from_vec::<VERSION>(label, key.to_vec());
+        let query = "UPDATE storage SET data = JSON_REMOVE(data, JSON_UNQUOTE(JSON_SEARCH(data, 'one', ?))) WHERE storage_key = ? AND version = ?";
+        self.conn.raw_query(|conn| {
+            sql_query(query)
+                .bind::<diesel::sql_types::Binary, _>(&value)
+                .bind::<diesel::sql_types::Binary, _>(&storage_key)
+                .bind::<diesel::sql_types::Text, _>(&VERSION.to_string())
+                .execute(conn)
+        });
+
+        Ok(())
+    }
+
+    pub fn read<const VERSION: u16, V: Entity<VERSION>>(
+        &self,
+        label: &[u8],
+        key: &[u8],
+    ) -> Result<Option<V>, <Self as StorageProvider<CURRENT_VERSION>>::Error> {
```

Comment:
It feels a little off that we do deserialization here, using a generic, but the `write` method just takes in `&[u8]`.

If we left serialization out of the base keystore methods it would also unlock us using the much more efficient `tls_codec` where it's available.